### PR TITLE
Add support for `dict` methods with `Const`

### LINKED ISSUE #2585: Bug: Dict methods do not work on a `Const dict`
Trying to use dict methods on `Const dict`  result in a `Internal Compiler Error`.

### get()

```python
d: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
print(d.get("a"))
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
.
.
.
AssertFailed: is_a<T>(*f)
```

### keys()
```python
d: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
print(d.keys())
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
.
.
.
AssertFailed: is_a<T>(*f)
```

The error is the same for other available methods.

### PR DESCRIPTION
**Pull request by kmr-srbh**
Fixes #2585 

Support for following attributes was implemented:
- `dict.get` - fetches the required value
- `dict.pop` - throws a `SemanticError` stating that changes cannot be made to a `const dict` 
- `dict.keys` - fetches the list of keys
- `dict.values` - fetches the list of values

### get()
```python
d: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
print(d.get("a"))
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
1
```

### pop()
```python
d: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
print(d.pop("a"))
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
semantic error: cannot pop elements from a const dict
 --> ./examples/example.py:8:7
  |
8 | print(d.pop("a"))
  |       ^^^^^^^^^^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lcompilers/lpython/issues.
```
### keys()
```python
d: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
print(d.keys())
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
['c', 'a', 'b']
```

### values()
```python
d: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
print(d.values())
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
[3, 1, 2]
```

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
@kmr-srbh  What issue does this PR fix?

File: src/lpython/semantics/python_attribute_eval.h [hunk 1]
```diff
@@ -331,7 +331,9 @@ struct AttributeHandler {
             throw SemanticError("'get' takes atleast 1 and atmost 2 arguments",
                     loc);
         }
-        ASR::ttype_t *type = ASRUtils::expr_type(s);
+        ASR::ttype_t *type = ASRUtils::type_to_str(ASRUtils::expr_type(s)) == "dict const"
+                                 ? ASRUtils::get_contained_type(ASRUtils::expr_type(s))
+                                 : ASRUtils::expr_type(s);
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_attribute_eval.h:336]: This does not seem to be the correct approach. See https://github.com/lcompilers/lpython/pull/2570#discussion_r1515196094.

If you want to check for `const` in type, use `ASR::is_a<ASR::Const_t>()`. 
If you want to get past the const and get to the actual type, use `ASRUtils::type_get_past_const()`.

Comment (kmr-srbh) [reply src/lpython/semantics/python_attribute_eval.h:336]: @Shaikh-Ubaid I had started with the above approach, but then switched as it threw this error every time:
image

**Comment by kmr-srbh** [comment]
> @kmr-srbh What issue does this PR fix?

I did not create a separate issue for this as I had worked out a fix. The main problem was that accessing values inside `Const` was not being handled when handling subscript indices. I had worked out #2560 for that. 

In the testing process I found that the dictionary methods were still failing. This PR adds support for using dictionary methods on a `Const dict`.

File: src/lpython/semantics/python_attribute_eval.h (hunk 1 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_attribute_eval.h:336]: The problem is resolved. My bad, I had missed the pointer asterisk. I am pushing the changes stated by you. :+1: Thank you very much!

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -1540,8 +1539,10 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void visit_DictItem(const ASR::DictItem_t& x) {
-        ASR::Dict_t* dict_type = ASR::down_cast<ASR::Dict_t>(
-                                    ASRUtils::expr_type(x.m_a));
+        ASR::Dict_t *dict_type = ASR::is_a<ASR::Const_t>(*ASRUtils::expr_type(x.m_a))
+                                     ? ASR::down_cast<ASR::Dict_t>(ASRUtils::type_get_past_const(ASRUtils::expr_type(x.m_a)))
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_llvm.cpp:1543]: `type_get_past_const()` already has a check `ASR::is_a<ASR::Const_t>()`, so the check for `ASR::is_a<ASR::Const_t>` here is redundant and thus not needed.

Comment (Shaikh-Ubaid) [reply src/libasr/codegen/asr_to_llvm.cpp:1543]: Similarly update others.

**Comment by Shaikh-Ubaid** [comment]
> > @kmr-srbh What issue does this PR fix?
> 
> I did not create a separate issue for this as I had worked out a fix.

Why don't you create an issue first and list all the bugs related to `Dict` and `Dict` operations that you discovered? Then in this PR specify exactly which bugs you have fixed (or are targeting to fix).

That would help highlight the contributions made. Thanks!

**Comment by Shaikh-Ubaid** [comment]
Also, you do not need to share the output as an image. Instead, you can simply share it in markdown as:

```python
(lp) ubaid@ubaids-MacBook-Pro lpython % cat examples/expr2.py  
from lpython import i32

def main0():
    x: i32
    x = (2+3)*5
    print(x)

if __name__ == "__main__":
   main0()
```
```console
(lp) ubaid@ubaids-MacBook-Pro lpython % lpython examples/expr2.py 
25
```
Sharing in text format as above is helpful as you can also search for some keyword in your code later.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 2 above)
Comment (kmr-srbh) [reply src/libasr/codegen/asr_to_llvm.cpp:1543]: I did not get what you meant @Shaikh-Ubaid. Can you please show the changes? How will I know if the dictionary is not `Const` if I don't check the type? Please guide.

**Comment by Thirumalai-Shaktivel** [comment]
Please mark this PR ready for review once it is ready.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 2 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:1543]: I think, @Shaikh-Ubaid, suggested to just do:
```c++
ASR::Dict_t *dict_type = ASR::down_cast<ASR::Dict_t>(ASRUtils::type_get_past_const(ASRUtils::expr_type(x.m_a)));
```

Comment (kmr-srbh) [reply src/libasr/codegen/asr_to_llvm.cpp:1543]: So `ASRUtils::type_get_pas_const()` returns the same passed value if the value is not `Const`. Got it @Thirumalai-Shaktivel! :+1:

Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:1543]: Yup!

**Comment by kmr-srbh** [comment]
I am working on the tests, in the meantime I request you to review this PR.

**Comment by kmr-srbh** [comment]
I am writing tests for this change and encountered this error:
```console
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store %dict* %const_dict, %dict* %d_float, align 8
 %dict = type { i32, i32, i32, %key_value*, i8*, i1 }Stored value type does not match pointer operand type!
  store %dict.1* %const_dict298, %dict.1* %d_int, align 8
 %dict.1 = type { i32, i32, i32, %key_value.0*, i8*, i1 }
```

My code is this:
```python
from lpython import i32, f64, str, dict, Const

def test_dict_const():
    d_int: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}

    assert d_int.get("a") == 1
    assert d_int.keys() == ["c", "a", "b"]
    assert d_int.values() == [3, 1, 2]

    d_float: Const[dict[str, f64]] = {"a": 1.0, "b": 2.0, "c": 3.0}
    assert d_float.get("a") == 1.0
    assert d_float.keys() == ["c", "a", "b"]
    assert d_float.values() == [3.0, 1.0, 2.0]

test_dict_const()
```

The assertions do not fail and the code runs normally when not enclosed within the function. Please look into it.

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid @Thirumalai-Shaktivel  This PR is now ready.

File: src/lpython/semantics/python_attribute_eval.h [hunk 3]
```diff
@@ -68,7 +68,7 @@ struct AttributeHandler {
     ASR::asr_t* get_attribute(ASR::expr_t *e, std::string attr_name,
             Allocator &al, const Location &loc, Vec<ASR::expr_t*> &args, diag::Diagnostics &diag) {
         ASR::ttype_t *type = ASRUtils::expr_type(e);
-        std::string class_name = get_type_name(type);
+        std::string class_name = ASR::is_a<ASR::Const_t>(*type) ? "dict" : get_type_name(type);
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_attribute_eval.h:71]:
```suggestion
        std::string class_name = get_type_name(ASRUtils::type_get_past_const(type));
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 4]
```diff
@@ -2790,14 +2790,17 @@ namespace DictValues {
...
+                                  : ASRUtils::expr_type(x.m_args[0]);
+    ASRUtils::require_impl(ASR::is_a<ASR::Dict_t>(*dict_type),
         "Argument to dict.values must be of dict type",
         x.base.base.loc, diagnostics);
     ASRUtils::require_impl(ASR::is_a<ASR::List_t>(*x.m_type) &&
-        ASRUtils::check_equal_type(ASRUtils::get_contained_type(x.m_type),
-        ASRUtils::get_contained_type(ASRUtils::expr_type(x.m_args[0]), 1)),
-        "Return type of dict.values must be of list of dict value element type",
-        x.base.base.loc, diagnostics);
+                               ASRUtils::check_equal_type(ASRUtils::get_contained_type(x.m_type),
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:2800]: Please remove all the formatting changes in this PR.
We limit the source code to 80 characters per line (more or less 10 characters), so do follow that style.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 5]
```diff
@@ -2790,14 +2790,17 @@ namespace DictValues {
 static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x, diag::Diagnostics& diagnostics) {
     ASRUtils::require_impl(x.n_args == 1, "Call to dict.values must have no argument",
         x.base.base.loc, diagnostics);
-    ASRUtils::require_impl(ASR::is_a<ASR::Dict_t>(*ASRUtils::expr_type(x.m_args[0])),
+    ASR::ttype_t *dict_type = ASR::is_a<ASR::Const_t>(*ASRUtils::expr_type(x.m_args[0]))
+                                  ? ASRUtils::type_get_past_const(ASRUtils::expr_type(x.m_args[0]))
+                                  : ASRUtils::expr_type(x.m_args[0]);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:2795]:
```suggestion
    ASR::ttype_t *dict_type = ASRUtils::type_get_past_const(ASRUtils::expr_type(x.m_args[0]));
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 6]
```diff
@@ -1540,8 +1539,8 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void visit_DictItem(const ASR::DictItem_t& x) {
-        ASR::Dict_t* dict_type = ASR::down_cast<ASR::Dict_t>(
-                                    ASRUtils::expr_type(x.m_a));
+        ASR::Dict_t *dict_type = ASR::down_cast<ASR::Dict_t>(ASRUtils::type_get_past_const(ASRUtils::expr_type(x.m_a)));
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:1542]: Limit to 80 characters per line.

File: tests/errors/test_dict_const.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from lpython import i32, f64, str, dict, Const
+
+
+def test_dict_const():
+    d: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
+    print(d.pop("a"))
+    i: i32 = d.pop("a")
```
Comment (Thirumalai-Shaktivel) [inline tests/errors/test_dict_const.py:7]: Add this tests to tests.toml and do `./run_tests.py -u`

File: src/libasr/pass/intrinsic_function_registry.h (hunk 4 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:2800]: I am sorry for this. The IDE did this on it's own.

File: tests/errors/test_dict_const.py (hunk 7 above)
Comment (kmr-srbh) [reply tests/errors/test_dict_const.py:7]: Hmm..so we need to do this manually? I thought `./run_tests.py -u` does this automatically. How about adding this information to the docs?

Comment (Thirumalai-Shaktivel) [reply tests/errors/test_dict_const.py:7]: It was present in the README.md, I think someone has remove the tests section in README.md

Comment (kmr-srbh) [reply tests/errors/test_dict_const.py:7]: It was me only @Thirumalai-Shaktivel . @certik suggested to keep it short in the README and redirect user to the installation docs for more. The testing section is there.

Comment (Thirumalai-Shaktivel) [reply tests/errors/test_dict_const.py:7]: Yes, understood. Now, it looks neat.
Thanks for doing it.

Comment (Thirumalai-Shaktivel) [reply tests/errors/test_dict_const.py:7]: > How about adding this information to the docs?

Yes, let's add it.

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid The same error here too: 
```console
code generation error: asr_to_llvm: module failed verification. Error:
Stored value type does not match pointer operand type!
  store %dict* %const_dict, %dict* %CONST_DICTIONARY_FLOAT, align 8
 %dict = type { i32, i32, i32, %key_value*, i8*, i1 }Stored value type does not match pointer operand type!
  store %dict.1* %const_dict298, %dict.1* %CONST_DICTIONARY_INTEGR, align 8
 %dict.1 = type { i32, i32, i32, %key_value.0*, i8*, i1 }
```

**Comment by Shaikh-Ubaid** [comment]
>@Shaikh-Ubaid The same error here too:

I believe the fix could/would be similar to the one we did for https://github.com/lcompilers/lpython/pull/2570. I will look into it tomorrow. Untill then check https://github.com/lcompilers/lpython/pull/2570/commits/0b9b94fd9dffad08adaec9f13a3bf7bc47f9e769 and see if you can figure out a fix along similar lines.

**Comment by kmr-srbh** [comment]
> Untill then check https://github.com/lcompilers/lpython/commit/0b9b94fd9dffad08adaec9f13a3bf7bc47f9e769 and see if you can figure out a fix along similar lines.

I am trying @Shaikh-Ubaid.

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid could you please look into the above error. I could not devote much time to work a fix for it. This is main blocker in this PR and #2579 .

**Comment by Thirumalai-Shaktivel** [comment]
Marking this PR as draft for now.

**Comment by Shaikh-Ubaid** [comment]
`Const` ASR node was recently removed. @kmr-srbh please, can you check if the examples shared in the PR description and the examples added in the integration tests in this PR work with the latest main? Can you share each example that you check and share its output here?

**Comment by kmr-srbh** [comment]
> `Const` ASR node was recently removed. @kmr-srbh please, can you check if the examples shared in the PR description and the examples added in the integration tests in this PR work with the latest main? Can you share each example that you check and share its output here?

I will the share the output. I think this should get back to working with a few small changes. The changes we had previously made to handle const list (merged) works. I think we will have to replace the `ASR::is_a<ASR::Const_t>` with `ASRUtils::is_const()`. The call to `ASRUtils::type_get_past_const()` must be removed too. I will look into this PR and make the necessary changes to get this working.

Thanks a lot for looking into this @Shaikh-Ubaid!

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid the PR is ready. Here is an example that works:
```python
CONST_DICT: Const[dict[str, i32]] = {"a": 1, "b": 2, "c": 3}
print(CONST_DICT.get("a"))
print(CONST_DICT.keys())
print(CONST_DICT.values())

# print(CONST_DICT.pop("a"))
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
1
['a', 'b', 'c']
[1, 2, 3]
```

Uncommenting the last line throws the required error:
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
semantic error: cannot pop elements from a const dict
 --> ./examples/example.py:6:7
  |
6 | print(CONST_DICT.pop("a"))
  |       ^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear or confusing messages as bugs at
https://github.com/lcompilers/lpython/issues.
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 8]
```diff
@@ -1524,8 +1524,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void visit_ListItem(const ASR::ListItem_t& x) {
-        ASR::ttype_t* el_type = ASRUtils::get_contained_type(
-                                        ASRUtils::expr_type(x.m_a));
+        ASR::ttype_t *el_type = ASRUtils::get_contained_type(ASRUtils::expr_type(x.m_a));
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_llvm.cpp:1527]: Let's not do this formatting change. It passes the `80` character limit in editors/terminal. The previous one seems better.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 9]
```diff
@@ -1541,8 +1540,8 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void visit_DictItem(const ASR::DictItem_t& x) {
-        ASR::Dict_t* dict_type = ASR::down_cast<ASR::Dict_t>(
-                                    ASRUtils::expr_type(x.m_a));
+        ASR::Dict_t *dict_type = ASR::down_cast<ASR::Dict_t>(ASRUtils::expr_type(x.m_a));
+
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_llvm.cpp:1544]: Similarly, please revert.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 10]
```diff
@@ -1846,8 +1845,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void generate_DictElems(ASR::expr_t* m_arg, bool key_or_value) {
-        ASR::Dict_t* dict_type = ASR::down_cast<ASR::Dict_t>(
-                                    ASRUtils::expr_type(m_arg));
+        ASR::Dict_t *dict_type = ASR::down_cast<ASR::Dict_t>(ASRUtils::expr_type(m_arg));
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_llvm.cpp:1848]: Similarly, please revert.

File: integration_tests/test_const_dict.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    assert CONST_DICTIONARY_INTEGR.get("a") == 1
+    assert CONST_DICTIONARY_INTEGR.keys() == ["c", "a", "b"]
+    assert CONST_DICTIONARY_INTEGR.values() == [3, 1, 2]
+
+    CONST_DICTIONARY_FLOAT: Const[dict[str, f64]] = {"a": 1.0, "b": 2.0, "c": 3.0}
+
+    assert CONST_DICTIONARY_FLOAT.get("a") == 1.0
+    assert CONST_DICTIONARY_FLOAT.keys() == ['a', 'b', 'c']
+    assert CONST_DICTIONARY_FLOAT.values() == [1.0, 2.0, 3.0]
+
```
Comment (Shaikh-Ubaid) [inline integration_tests/test_const_dict.py:16]: Please print values before the asserts. Also, please enable this test in `integration_tests/CMakeLists.txt`.

File: tests/tests.toml [hunk 12]
```diff
@@ -497,6 +497,10 @@ asr = true
 filename = "../integration_tests/global_syms_01.py"
 asr = true
 
+[[test]]
+filename = "../integration_tests/test_const_dict.py"
+asr = true
+
```
Comment (Shaikh-Ubaid) [inline tests/tests.toml:503]: I think we need not test this here. Testing it as an `integration_test` is sufficient.

Comment (Shaikh-Ubaid) [reply tests/tests.toml:503]: After you remove this from here, please also remove the generated reference files.

**Review by Shaikh-Ubaid — COMMENTED** [review]
Thanks for this! I shared some comments. I will review again after these are resolved.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 13]
```diff
@@ -1524,7 +1524,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void visit_ListItem(const ASR::ListItem_t& x) {
-        ASR::ttype_t* el_type = ASRUtils::get_contained_type(
+        ASR::ttype_t *el_type = ASRUtils::get_contained_type(
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_llvm.cpp:1527]: Please revert the style changes.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 14]
```diff
@@ -1846,8 +1847,8 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void generate_DictElems(ASR::expr_t* m_arg, bool key_or_value) {
-        ASR::Dict_t* dict_type = ASR::down_cast<ASR::Dict_t>(
-                                    ASRUtils::expr_type(m_arg));
+        ASR::Dict_t *dict_type = ASR::down_cast<ASR::Dict_t>(
+                                            ASRUtils::expr_type(m_arg));
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_llvm.cpp:1851]: Please revert.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 13 above)
Comment (kmr-srbh) [reply src/libasr/codegen/asr_to_llvm.cpp:1527]: Thank you very much for bringing this to notice @Shaikh-Ubaid! We really do not need these changes as there are no real changes. :+1: I forgot to revert them when reverting the previous changes. I am doing this now.

**Review by Shaikh-Ubaid — APPROVED** [review]
It looks good. Thanks.

**Pull request merged**